[PM-14818] Adjust instructions for running EF integration tests (#457)
bitwarden/contributing-docs: docs/contributing/testing/database/index.md

# Database Integration Testing

> Since Bitwarden has multiple database options, testing them all automatically is incredibly
> important so that we don't have to have a full QA process for each database type. This is where the
> `Infrastructure.IntegrationTest` project comes in, allowing the setup of tests similarly to how the
> databases are consumed in the applications through their common interfaces, generally an
> `I*Repository`. These tests are automatically executed on all supported databases in our testing
> pipeline.

Reviewer: I find this a bit hard to follow and is probably packing too much info into the opening paragraph. How about:

```suggestion
Since Bitwarden has multiple database options, automated testing is extremely important to avoid time-consuming manual testing for each database type. The `Infrastructure.IntegrationTest` project allows you to write tests that are automatically executed on all supported databases in our testing pipeline.
```

## Creating a new test

> To create a new database test just add the `[DatabaseTheory]` and `[DatabaseData]` attributes to
> test. Now, in the parameters of the test you are able to "inject" any repository layer services
> directly into the test. The test will run for every database that is
> [configured in the current environment](#configure-the-tests). Since you inject the interface of the
> service, some runs of the test will use the Dapper based repository implementation targeting
> Microsoft SQL Server and others will use the Entity Framework Core based implementations, which we
> use for MySql, Postgres, and SQLite.

Reviewer: ```suggestion
To create a new database test, add the `[DatabaseTheory]` and `[DatabaseData]` attributes to
test. Then, use the parameters of the test to inject any repository layer services
you need. The test will run for every database that is
[configured in the current environment](#configure-the-tests). Since you inject the interface of the
service, some runs of the test will use the Dapper-based repository implementation targeting
Microsoft SQL Server and others will use the Entity Framework Core based implementations (which we
use for MySql, Postgres, and SQLite).
```

> The goal of database tests is to test the business logic that is encapsulated in a given method. If
> the stored procedure in SQL Server then calls another procedure to update the
> `User.AccountRevisionDate` then the same EF implementation should do that as well. By running the
> test against all variants we are ensuring all the variants are feature-equal. Locally, you may want

Reviewer: ```suggestion
The goal of database tests is to test the business logic that is encapsulated in a given method. For example, if
a stored procedure in SQL Server calls another procedure to update the
`User.AccountRevisionDate` then the corresponding EF implementation should do that as well. By running the
test against all variants, we are ensuring all the variants are feature-equal.
```

> to only run the SQL Server tests along with one EF implementation; SQLite is often the easiest in
> that situation. This may work well for a very long time and save you some time overall but there are
> differences between the EF database providers such that you will one day get errors in the CI
> pipeline.

Reviewer: I'm not sure what the doc is recommending here. Should we use SQLite because it's easy? Or all 3 unsupported-supported EF databases to be comprehensive?

Given that it runs in the pipeline I think we can recommend the easier path for local development.

```suggestion
to only run the SQL Server tests along with one EF implementation; SQLite is often the easiest in
that regard. The other supported EF database providers will still run in the pipeline to catch any
differences between them.
```

## Configure the databases

> The databases are expected to have the latest migrations applied.

Reviewer: Link to your instructions on this?

Author: They are already linked below for running the tests locally. I don't really think this section is offering anything and opted to remove it instead.

## Configure the tests

> The tests are configured through
> [.NET Configuration](https://learn.microsoft.com/en-us/dotnet/core/extensions/configuration) in the
> order they are applied: user secrets, environment variables prefixed with `BW_TEST_`, and command
> line args.

Reviewer: ```suggestion
The tests are configured through
[.NET Configuration](https://learn.microsoft.com/en-us/dotnet/core/extensions/configuration). They are applied in the following order: user secrets, environment variables prefixed with `BW_TEST_`, and command
line args.
```
